### PR TITLE
Event menu rewrite

### LINKED ISSUE #68408: Admin triggering events will put it on a 10 second timer ("Random event triggerin in [x]"), and ask "if you'd like to alert the crew" after it's already triggered
- Admin trigger an event
- It puts it on a 10 second random event "trigger" queue, instead of it being instant
- After the event actually triggers, it asks you if you want to alert the crew or cancel

### PR DESCRIPTION
**Pull request by Profakos**
## About The Pull Request

kép

This PR does the following:

- Force event menu uses tgUI.
- Arranged events into categories, and added a little description to each. The descriptions appear as tooltips when you hover over the Trigger button.
- Rewrote how "Announce to crew?" works. It no longer pops up a panel after the event has been already announced. Instead, the admins select it via a checkbox, and the result is passed through an optional argument.
- announceChance's comment is tweaked a bit to reflect how it actually works at the moment.
- Moved rpgtitles to wizard events, where it belongs.
- Fake Virus and Electric Storms show up to observers, as I believe they are not as common as Space Dust or Camera Failure, and should be cancelable.

Potential issues:
~~This only solves half of #68408, I don't think admin triggering having a timer and a cancel button is a big issue, as it allows other admins to overrule you if needed, but if i is, I will try to fix it within this PR.~~

Fixes #68408. Events now spawn immediately, and the the announceChance is overwritten before it begins.

My choices for categories and descriptions might not be the best, feedback would be appreciated.

## Why It's Good For The Game

The old spawn menu was completely unorganized, and you could only search using the browser search tool. I believe a built in search bar helps with this issue a bit. I also believe that organizing the events into categories, and adding descriptions will help with newer admins who might not be familiar with all events.

## Changelog

:cl:
refactor: The Force Event UI has been refactored
refactor: Events now have categories and descriptions
refactor: Admin triggered events happen immediately
balance: Fake Virus and Electric Storms are shown to admins, making them cancelable
/:cl:

### CODE REVIEW

File: code/__DEFINES/events.dm [hunk 1]
```diff
@@ -7,3 +7,15 @@
...
+#define EVENT_CAT_ANOMALIES "Anomalies"
+#define EVENT_CAT_BUREAUCRATIC "Bureaucratic"
+#define EVENT_CAT_ENGINEERING "Engineering"
+#define EVENT_CAT_ENTITIES "Entities"
+#define EVENT_CAT_FRIENDLY "Friendly"
+#define EVENT_CAT_HEALTH "Health"
+#define EVENT_CAT_HOLIDAY "Holiday"
+#define EVENT_CAT_INVASION "Invasion"
+#define EVENT_CAT_SPACE "Space Threats"
+#define EVENT_CAT_WIZARD "Wizard"
```
Comment (Mothblocks) [inline code/__DEFINES/events.dm:21]: 1. Doc-comment all these (///)
2. CATEGORY, not CAT

File: code/modules/admin/force_event.dm [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    switch(action)
+        if("forceevent")
+            var/announce_event = params["announce"]
+            var/string_path = params["type"]
+            if(!string_path)
+                return
+            var/event_to_run_type = text2path(string_path)
+            if(!event_to_run_type)
+                return
+            var/datum/round_event_control/E = locate(event_to_run_type) in SSevents.control
```
Comment (Mothblocks) [inline code/modules/admin/force_event.dm:61]: Write out the full name

File: code/modules/admin/force_event.dm [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+            if(E)
```
Comment (Mothblocks) [inline code/modules/admin/force_event.dm:62]: Early-return

File: code/modules/admin/force_event.dm [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+                E.admin_setup(usr)
+                var/datum/round_event/event = E.runEvent(announce_chance_override = announce_event ? 100 : 0)
```
Comment (Mothblocks) [inline code/modules/admin/force_event.dm:64]: What is 100 here? Is it meant to be 10 SECONDS?

File: code/modules/admin/force_event.dm [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+                if(event.cancel_event)
```
Comment (Mothblocks) [inline code/modules/admin/force_event.dm:65]: When is this set? An admin triggering an event should trigger the event, not delay it.

File: tgui/packages/tgui/interfaces/ForceEvent.js [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+import { Box, Stack, Button, Icon, Input, Section, Table } from '../components';
+import { Window } from '../layouts';
+import { flow } from 'common/fp';
+import { filter, sortBy } from 'common/collections';
+
+export const ForceEvent = (props, context) => {
+  return (
+    <Window title="Force Event" width={450} height={450}>
+      <Window.Content scrollable>
+        <Stack fill vertical>
```
Comment (Mothblocks) [inline tgui/packages/tgui/interfaces/ForceEvent.js:11]: Does this need to be a stack?

File: tgui/packages/tgui/interfaces/ForceEvent.js [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+  const [announce, setAnnounce] = useLocalState(context, 'announce', true);
+
+  return (
+    <Stack vertical fill>
+      <Stack.Item>
+        <Button.Checkbox
+          fluid
+          checked={announce}
+          onClick={() => setAnnounce(!announce)}>
+          Announce event to the crew {announce}
```
Comment (Mothblocks) [inline tgui/packages/tgui/interfaces/ForceEvent.js:66]: What is `{announce}` here for? Is the checkbox on its own not okay?

**Comment by Profakos** [comment]
I am confused by the linter, its the mandatory prettierx run that makes my JS files have space indentations...

File: tgui/packages/tgui/interfaces/ForceEvent.js [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+  );
+};
+
+export const EventOptionsPanel = (props, context) => {
+  const { data } = useBackend(context);
+
+  const [announce, setAnnounce] = useLocalState(context, 'announce', true);
+
+  return (
+    <Stack vertical fill>
```
Comment (Mothblocks) [inline tgui/packages/tgui/interfaces/ForceEvent.js:60]: Do you need a stack for a single item?

**Comment by Profakos** [comment]
Ah, I do not know how space indentation appeared in the .dm ... I shall fix it posthaste.

File: tgui/packages/tgui/interfaces/ForceEvent.js [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+    if (nameA > nameB) {
+      return 1;
+    }
+    return 0;
+  };
+
+  return (
+    <Section>
+      <Stack vertical fill>
+        {categories.sort(nameSorter).map((category) => (
```
Comment (Mothblocks) [inline tgui/packages/tgui/interfaces/ForceEvent.js:93]: You are mutating `categories`, which mutates `data`, which is (as far as I am aware), not allowed.

Use `sortBy` instead.

File: tgui/packages/tgui/interfaces/ForceEvent.js [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+          <Stack.Item mt={0.2}>
```
Comment (Mothblocks) [inline tgui/packages/tgui/interfaces/ForceEvent.js:94]: This needs a key.

```suggestion
          <Stack.Item mt={0.2} key={category.name}>
```

**Review by Mothblocks — CHANGES_REQUESTED** [review]

**Comment by Mothblocks** [comment]
> I don't think admin triggering having a timer and a cancel button is a big issue

the comedic timing is

File: code/modules/admin/force_event.dm (hunk 4 above)
Comment (Profakos) [reply code/modules/admin/force_event.dm]: Probability percentage. Overrides announceChance, which is used as such:```	if(activeFor == announceWhen && prob(announceChance))
		processing = FALSE
		announce(FALSE)
		processing = TRUE```

File: code/modules/admin/force_event.dm (hunk 5 above)
Comment (Profakos) [reply code/modules/admin/force_event.dm]: When a different admin clicks on the cancel link href shown here https://github.com/tgstation/tgstation/blob/f0ec14736c3fa1a83229885976795ded44c67cc0/code/modules/events/_event.dm#L112

File: tgui/packages/tgui/interfaces/ForceEvent.js (hunk 7 above)
Comment (Profakos) [reply tgui/packages/tgui/interfaces/ForceEvent.js:66]: Leftover from testing, when it was not working properly due to setting a different default value in localState by accident. Removing it asap.

File: code/modules/admin/force_event.dm (hunk 5 above)
Comment (Profakos) [reply code/modules/admin/force_event.dm]: Ah I also see an error I did: I wrote out the full name of the E variable just here, and forgot about the rest.

**Comment by Profakos** [comment]
I shall make them trigger automatically in that case.

**Comment by Profakos** [comment]
Lessons learned: I should run the linters locally before pushing.

File: tgui/packages/tgui/interfaces/ForceEvent.js (hunk 10 above)
Comment (Profakos) [reply tgui/packages/tgui/interfaces/ForceEvent.js]: Key has been added.

File: tgui/packages/tgui/interfaces/ForceEvent.js (hunk 9 above)
Comment (Profakos) [reply tgui/packages/tgui/interfaces/ForceEvent.js]: I have swapped to sortBy.

File: tgui/packages/tgui/interfaces/ForceEvent.js (hunk 8 above)
Comment (Profakos) [reply tgui/packages/tgui/interfaces/ForceEvent.js]: Ah, I have been overzealous. I have changed it.

File: tgui/packages/tgui/interfaces/ForceEvent.js (hunk 6 above)
Comment (Profakos) [reply tgui/packages/tgui/interfaces/ForceEvent.js]: Turns out, it does not, thank you.

File: code/modules/admin/force_event.dm (hunk 5 above)
Comment (Profakos) [reply code/modules/admin/force_event.dm]: I have removed the cancelability and the delay from admin triggers as requested.

File: code/modules/admin/force_event.dm (hunk 3 above)
Comment (Profakos) [reply code/modules/admin/force_event.dm]: Early return has been added.

File: code/modules/admin/force_event.dm (hunk 2 above)
Comment (Profakos) [reply code/modules/admin/force_event.dm]: Done.

File: code/__DEFINES/events.dm (hunk 1 above)
Comment (Profakos) [reply code/__DEFINES/events.dm]: All done.

File: code/modules/admin/force_event.dm (hunk 4 above)
Comment (Profakos) [reply code/modules/admin/force_event.dm]: I could alternatively replace announce_chance_override with a three stat define, saying ANNOUNCE_CHANCE_USE_DEFAULT, ANNOUNCE_CHANCE_SILENT, and ANNOUNCE_CHANCE_ALWAY_ANNOUNCE

Comment (Profakos) [reply code/modules/admin/force_event.dm]: In the meanwhile, I have demagiced the numbers with two vars, to better explain in code what happens there.

**Comment by Cheshify** [comment]
This is going to be really nice, makes spawning rods at the station so much easier!

File: tgui/packages/tgui/interfaces/ForceEvent.js [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+};
+
+export const EventContent = (props, context) => {
+  const { data } = useBackend(context);
+
+  const categories = Object.values(data.categories);
+  const sortCategories = sortBy((category) => category.name);
+
+  return (
+    <Section>
```
Comment (stylemistake) [inline tgui/packages/tgui/interfaces/ForceEvent.js:68]: I don't think you need this section, and you don't need a vertical stack there either. In the map, directly return EventList elements.

Comment (Profakos) [reply tgui/packages/tgui/interfaces/ForceEvent.js]: Thank you, I have done so.

**Comment by Profakos** [comment]
I realize an oversight i have made. It looks like instead of admin_state, i should have made a brand new fun_state, as the original code was only locked  behind having R_FUN, so i accidentaly decreased the number of people who can actually use the interface. I shall fix this soon (unless if someone tells me this is better this way, that is)

**Comment by Profakos** [comment]
There, I have added the Fun State, the state for having fun times. Now once again you only need to have fun to trigger events! :y

**Review by stylemistake — APPROVED** [review]

**Comment by Profakos** [comment]
@Mothblocks I would like to ask for the re-review, it has been some time, and I think I have resolved all the raised issues...

**Comment by Mooshimi** [comment]
> @Mothblocks I would like to ask for the re-review, it has been some time, and I think I have resolved all the raised issues...

Mothblocks is on vacation currently. 

<details><summary> Quoth the Mothblocks:</summary>

</details>

**Comment by Profakos** [comment]
Ah, thank you. I will be patient.

**Comment by 13spacemen** [comment]
@Profakos Very cool and necessary PR
Small request, please add a separate "Cancel" button next to "Trigger"
Currently, with the non TGUI menu, if you click the event name button twice quickly, instead of doing 2 events it will do 1 event and then cancel it, very annoying since there's no indication that clicking twice cancels

**Comment by Profakos** [comment]
Do you mean that you want the trigger button to replaced by Cancel after you hit trigger and before the event finishes setting up?

**Comment by 13spacemen** [comment]
@Profakos 
untitled

I mean having a separate Cancel button to cancel. Currently with the non-TGUI menu, if you click an event twice quickly, it will cancel the event instead of simply spawning 2 events.

**Comment by Profakos** [comment]
That is rather confusing. I would have to make the cancel button be disabled most of the time, except when an event is running... however the event has already begun at this time, as requested in the issue report, there is no delay, so it can not be cancelled anymore. 
I have to check if this double click behaviour is still present or not. It feels like a bug instead of a feature.

**Comment by Profakos** [comment]
I have checked, and this is no longer needed, as the code no longer sleeps, it will spawn an event for every click of the trigger button.
Maybe I should add a cooldown to this menu.

**Comment by Profakos** [comment]
Should i add a debounce to the trigger buttons, or is it good to go like this?

**Comment by Profakos** [comment]
Just a heads up, I will be on vacation from the 7th to the 13th, so if this gets merged and there are issues I won't be around to fix it, however I have asked a few of my coder friends to look into possible problems.

**Comment by optimumtact** [comment]
Me sending it anyway like a chad

**Comment by Profakos** [comment]
Thank you! Hoping there won't be any issues...

**Pull request merged**